cmd/abapEnvironmentCreateTag.go

```diff
@@ -0,0 +1,246 @@
+
+	for index := range item.tags {
+		err = createSingleTag(item, index, telemetryData, con, client, com)
+	}
+	return err
+}
+
+func createSingleTag(item CreateTagBacklog, index int, telemetryData *telemetry.CustomData, con abaputils.ConnectionDetailsHTTP, client piperhttp.Sender, com abaputils.Communication) (err error) {
+
+	requestBodyStruct := CreateTagBody{RepositoryName: item.repositoryName, CommitID: item.commitID, Tag: item.tags[index].tagName, Description: item.tags[index].tagDescription}
+	errorMessage := "Could not create tag " + requestBodyStruct.Tag + " for repository " + requestBodyStruct.RepositoryName + " with commitID " + requestBodyStruct.CommitID
+	requestBodyJson, err := json.Marshal(&requestBodyStruct)
+	if err != nil {
+		return err
+	}
+
+	log.Entry().Debugf("Request body: %s", requestBodyJson)
+	resp, err := abaputils.GetHTTPResponse("POST", con, requestBodyJson, client)
+	if err != nil {
+		err = abaputils.HandleHTTPError(resp, err, errorMessage, con)
+		return err
+	}
+	defer resp.Body.Close()
+
+	// Parse response
+	var body CreateTagResponse
+	var abapResp map[string]*json.RawMessage
+	bodyText, _ := ioutil.ReadAll(resp.Body)
+
+	json.Unmarshal(bodyText, &abapResp)
+	json.Unmarshal(*abapResp["d"], &body)
+
+	con.URL = con.Host + "/sap/opu/odata/sap/MANAGE_GIT_REPOSITORY/Pull(guid'" + body.UUID + "')"
+	err = checkStatus(con, client, com)
+
+	if err == nil {
+		log.Entry().Info("Created tag " + requestBodyStruct.Tag + " for repository " + requestBodyStruct.RepositoryName + " with commitID " + requestBodyStruct.CommitID)
+	} else {
+		log.Entry().Error("NOT created: Tag " + requestBodyStruct.Tag + " for repository " + requestBodyStruct.RepositoryName + " with commitID " + requestBodyStruct.CommitID)
+	}
+
+	return err
+}
+
+func checkStatus(con abaputils.ConnectionDetailsHTTP, client piperhttp.Sender, com abaputils.Communication) (err error) {
+	status := "R"
+	pollIntervall := com.GetPollIntervall()
+	count := 0
+	for {
+		count += 1
+		entity, _, err := abaputils.GetStatus("Could not create Tag", con, client)
+		if err != nil {
+			return err
+		}
+		status = entity.Status
+		if status != "R" {
+			if status == "E" {
+				err = errors.New("Could not create Tag")
+			}
+			return err
+		}
+		if count >= 200 {
+			return errors.New("Could not create Tag (Timeout)")
+		}
+		time.Sleep(pollIntervall)
+	}
+}
+
+func prepareBacklog(config *abapEnvironmentCreateTagOptions) (backlog []CreateTagBacklog, err error) {
+
+	descriptor, err := abaputils.ReadAddonDescriptor(config.Repositories)
+	if err != nil {
+		return
+	}
+	repos := descriptor.Repositories
+
+	for _, repo := range repos {
+
+		backlogInstance := CreateTagBacklog{repositoryName: repo.Name, commitID: repo.CommitID}
+		if repo.VersionYAML != "" {
+			tag := Tag{tagName: "v" + repo.VersionYAML, tagDescription: "Generated by the ABAP Environment Pipeline"}
```

Comment:
if the same repo is part of multiple addon products the same tag will be created multiple times, I hope this is no issue?

Reply:
The API does not yet return "does already exist". But once it does, it is possible to react accordingly in the piper step. I think it makes sense to only have one tag for a specific version